phoenix-core/src/main/java/org/apache/phoenix/coprocessor/GroupedAggregateRegionObserver.java

```diff
@@ -373,198 +389,253 @@ GroupByCache newCache(RegionCoprocessorEnvironment env, ImmutableBytesPtr tenant
+                        // value of s.next is false
+                        // since this is an indication of whether or not there
+                        // are more values after the
+                        // ones returned
+                        hasMore = delegate.nextRaw(kvs);
+                        if (!kvs.isEmpty()) {
+                            result.setKeyValues(kvs);
+                            key = TupleUtil.getConcatenatedValue(result, expressions);
+                            aggBoundary = currentKey != null && currentKey.compareTo(key) != 0;
+                            if (!aggBoundary) {
+                                aggregators.aggregate(rowAggregators, result);
+                                if (LOGGER.isDebugEnabled()) {
+                                    LOGGER.debug(LogUtil.addCustomAnnotations(
                                             "Row passed filters: " + kvs
-                                            + ", aggregated values: "
-                                            + Arrays.asList(rowAggregators),
+                                                    + ", aggregated values: "
+                                                    + Arrays.asList(rowAggregators),
                                             ScanUtil.getCustomAnnotations(scan)));
-                                    }
-                                    currentKey = key;
                                 }
+                                currentKey = key;
                             }
-                            atLimit = rowCount + countOffset >= limit;
-                            // Do rowCount + 1 b/c we don't have to wait for a complete
-                            // row in the case of a DISTINCT with a LIMIT
-                        } while (hasMore && !aggBoundary && !atLimit);
-                    }
-                } finally {
-                    if (acquiredLock) region.closeRegionOperation();
+                        }
+                        atLimit = rowCount + countOffset >= limit;
+                        // Do rowCount + 1 b/c we don't have to wait for a complete
+                        // row in the case of a DISTINCT with a LIMIT
+                        now = EnvironmentEdgeManager.currentTimeMillis();
+                    } while (hasMore && !aggBoundary && !atLimit && (now - startTime) < pageSizeInMs);
                 }
-
-                if (currentKey != null) {
-                    byte[] value = aggregators.toBytes(rowAggregators);
-                    KeyValue keyValue =
-                            KeyValueUtil.newKeyValue(currentKey.get(), currentKey.getOffset(),
+            } finally {
+                if (acquiredLock) region.closeRegionOperation();
+            }
+            if (hasMore && !aggBoundary && !atLimit && (now - startTime) >= pageSizeInMs) {
+                // Return a dummy result as we have processed a page worth of rows
```

Comment:
I thought we **want** to return a dummy result here too. Previously you had a call to `getDummyResult()` here but it was commented and I thought it should be uncommented. If we don't want to return the dummy result here then can you please explain why that is the case? We should remove these comments as well in that case.

Reply:
Yes, you are right. It was commented by mistake. I will add this back. Thank you for catching it.

Reply:
FYI ... This was missed in the patch I prepared for 4.x and thus I had to push an addendum for this one line change. The patch for master is correct.